[FAB-17145] Add test network to Fabric Samples

nikhil550 opened the pull request:
Addresses: FabricFAB-17147
New sample is meant to be long term replacement for BYFN. 
- New sample is simpler and more organized
- Bypasses tools container for org1 and org2
- Provides more options than the e2e flow of BYFN
- Provides support for using CA's in addition to cryptogen

The network is meant to be run on Fabric 2.0. To test the sample, you need to build the latest images from master locally. You also need to build the latest binaries for cryptogen and peer, and move then into the bin folder of your local fabric samples. You also need to copy the latest configuration files into the config file inside fabric samples. These are done when you clone the fabric samples, images, and binaries.

Signed-off-by: NIKHIL E GUPTA <negupta@us.ibm.com>

nikhil550 commented on test-network/network.sh:
```diff
@@ -0,0 +1,565 @@
+  # Create crypto material using Fabric CAs
+  if [ "$CRYPTO" == "Certificate Authorities" ]; then
+
+    fabric-ca-client version > /dev/null 2>&1
```
@mbwhite Added this section to download the fabric-ca-client if it is not available locally.

nikhil550 commented:
The script assumes that you followed the steps to install the sample and the binaries: https://hyperledger-fabric.readthedocs.io/en/latest/install.html

This downloads the binaries in a bin folder and the latest configuration files in a config folder. You can copy the binaries and configuration files to those locations to test.

nikhil550 commented on test-network/README.md:
```diff
@@ -0,0 +1,48 @@
+on those channels.
+
+Before you can deploy the test network, you need follow the instructions to 
+[Install the Samples, Binaries and Docker Images](https://hyperledger-fabric.readthedocs.io/en/latest/install.html) in the Hyperledger Fabric documentation.
```
Let folks now they need to install the binaries.

nikhil550 commented on test-network/network.sh:
```diff
@@ -0,0 +1,573 @@
+  ## Check if your have cloned the peer binaries and configuration files.
+  peer version > /dev/null 2>&1
+
+  if [[ $? -ne 0 || ! -d "../config" ]]; then
```
Let folks know and exit if the config files and binaries are missing.

nikhil550 commented:
> FYI - going to pull this locally and give it a test run :-)
> 
>     * network.sh - would be good to add `set -euo pipefail` at the top so any failures, eg missing commands stop the script immediately
> 
>     * I can't see if there's a 'clear config' command.  So for example it's not possible to force regeneration of the crypto material.
> 
>     * fabric-ca-client missing - doesn't give a warning
> 
>     * core.yaml missing?

@mbwhite on the points above
- Seems to encounter problems running `set -euo pipefail`
- I remove the certs every time I run /network.up. Leaving certs around could cause errors with left around artifacts. These artifacts are also cleared when you run ./network down
- Added command download fabric-ca-client if not found
- You need to download core.yaml with the binaries. Added checks to see if they are there. See above.

rameshthoomu commented:
Good work @nikhil550 
1) This script will fail when you build images and binaries. Config directory is missing. 2019-12-09 17:41:25.710 EST [main] InitCmd -> ERRO 001 Fatal error when initializing core config : FABRIC_CFG_PATH /Users/thoomu@us.ibm.com/gopath/src/github.com/hyperledger/fabric-samples/test-network/../config/ does not exist

nikhil550 commented:
> Good work @nikhil550
> 
>     1. This script will fail when you build images and binaries.
>        Config directory is missing. 2019-12-09 17:41:25.710 EST [main] InitCmd -> ERRO 001 Fatal error when initializing core config : FABRIC_CFG_PATH /Users/thoomu@us.ibm.com/gopath/src/github.com/hyperledger/fabric-samples/test-network/../config/ does not exist

Thanks @rameshthoomu 
Addressed with most recent changes above.

nikhil550 commented:
Added some changes in response to feedback from @rameshthoomu 
- Added the generate back to the add Org3 script and cleaned up the help text
- Fixed the path in the org3 script to point to the right binaries
- Added retry to the createChannel command to allow raft to elect leader 
- redid the joinchannel with retry to match other commands in the script

nikhil550 commented on test-network/README.md:
```diff
@@ -0,0 +1,50 @@
+Use the `./network.sh` script to stand up a simple Fabric test network. The
+network has two peer peer organizations with one peer each and a single node
+raft ordering service. You can also use the script to create channels, and deploy
+the fabcar chaincode on those channels. The test network is being introduced in
```
Per @lehors suggestion, added a sentence that this will be a long term replacement for BYFN.

lehors commented:
I'm going to ignore the CI error which is unrelated to this change and merge it.
